### DISCUSSION TITLE
`Failed to discover workspace` warning when opening a standalone file

### DISCUSSION THREAD

**Discussion by barraIhsan** [mrcjkb/rustaceanvim#525] [Q&A] [answered]
### Checked other resources

- [X] This is not a bug report or feature request (Please use [issues](https://github.com/mrcjkb/rustaceanvim/issues/new/choose) for those).
- [X] I have read the README.
- [X] I have searched for a solution using `:h rustaceanvim`.
- [X] I have searched [existing discussions](https://github.com/mrcjkb/rustaceanvim/discussions).
- [ ] I have searched the existing issues/discussions of other plugins related to this question (if applicable).
- [ ] I have searched the [existing rust-analyzer discussions](https://github.com/rust-lang/rust-analyzer/discussions) (if applicable).
- [X] I have added a very descriptive title to this question.

### Description

I'm trying to use `rustaceanvim` with a standalone/single-file rust.
However, I have this error, every single time I save (`:w`) the file:
![image](https://github.com/user-attachments/assets/0ad1e5e2-7864-40c1-bad7-89e4dcf26518)
Log:
```
ERROR rust_analyzer::main_loop: FetchWorkspaceError:
rust-analyzer failed to discover workspace
```

I thought that this feature is enabled by default on `rustaceanvim`?
```
        {standalone?}            (boolean)
                                                                                                                   Standalone file support (enabled by default).
                                                                                                                   Disabling it may improve rust-analyzer's startup time.
```
Or am I missing out on something?

### Example Code

```lua
-- Bootstrap lazy.nvim
local lazypath = vim.fn.stdpath 'data' .. '/lazy/lazy.nvim'
if not (vim.uv or vim.loop).fs_stat(lazypath) then
  local lazyrepo = 'https://github.com/folke/lazy.nvim.git'
  local out = vim.fn.system { 'git', 'clone', '--filter=blob:none', '--branch=stable', lazyrepo, lazypath }
  if vim.v.shell_error ~= 0 then
    error('Error cloning lazy.nvim:\n' .. out)
  end
end
vim.opt.rtp:prepend(lazypath)

require("lazy").setup({
  {
    "mrcjkb/rustaceanvim",
    version = "^5",
    lazy = false,
  }
})
```

### Neovim version (nvim -v)

v0.10.1

### Operating system/version

ArchWSL 5.15.153.1

### Output of :checkhealth rustaceanvim

```console
==============================================================================
rustaceanvim: require("rustaceanvim.health").check()

Checking for Lua dependencies ~
- WARNING dap not installed. Needed for debugging features [mfussenegger/nvim-dap](https://github.com/mfussenegger/nvim-dap)

Checking external dependencies ~
- OK rust-analyzer: found rust-analyzer 1.81.0 (eeb90cd 2024-09-04)
- OK Cargo: found cargo 1.81.0 (2dbb1af80 2024-08-20)
- OK rustc: found rustc 1.81.0 (eeb90cda1 2024-09-04)
- OK debug adapter: found codelldb 

Checking config ~
- OK No errors found in config.

Checking for conflicting plugins ~
- OK No conflicting plugins detected.

Checking for tree-sitter parser ~
- OK tree-sitter parser for Rust detected.
```

**Comment by mrcjkb** [comment, marked as answer]
Hey 👋 

That's an error reported by rust-analyzer.
I would consider it a bug, since rust-analyzer does have basic support for standalone files. At the very least it should be a reported as a warning, not an error (rustaceanvim only warns about rust-analyzer error status reports by default).

If it bothers you, you can disable rust-analyzer server status notifications by setting the `vim.g.rustaceanvim.server.status_notify_level` option to `false`.
However, then you won't be notified about any rust-analyzer errors.

**Reply by barraIhsan** [reply]
![image](https://github.com/user-attachments/assets/612a59c6-cb5d-425a-960e-d9a11684d227)
this also a bug from their end?

**Reply by barraIhsan** [reply]
> If it bothers you, you can disable rust-analyzer server status notifications by setting the vim.g.rustaceanvim.server.status_notify_level option to false.

uhm, because of that error, I cannot see the diagnostic provided by the linter `rust-analyzer`.

Inside cargo project:
![image](https://github.com/user-attachments/assets/2ef73f4e-7ada-45dc-8787-1f662a26f218)

Standalone file:
![image](https://github.com/user-attachments/assets/7006af34-e539-4635-9210-955297a09ecc)

**Reply by mrcjkb** [reply]
> ![image](https://private-user-images.githubusercontent.com/57800056/371503971-612a59c6-cb5d-425a-960e-d9a11684d227.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc0MzU5MjQsIm5iZiI6MTcyNzQzNTYyNCwicGF0aCI6Ii81NzgwMDA1Ni8zNzE1MDM5NzEtNjEyYTU5YzYtY2I1ZC00MjVhLTk2MGUtZDlhMTE2ODRkMjI3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI3VDExMTM0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWViOGZlNjVhYWI5YjlmMGExYWI2YTliMGJiYjViZjlhNzU4NmE1MjE2NmQ4YzA1MzcxNTU3MjJmN2Y4NGRiNTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.M0tw0qyVddLrOG8nRYQ3imvWyX3X3mqu50k6ml0VPaA) this also a bug from their end?

I don't know what that is, but it's coming from rust-analyzer.

**Reply by mrcjkb** [reply]
> > If it bothers you, you can disable rust-analyzer server status notifications by setting the vim.g.rustaceanvim.server.status_notify_level option to false.
> 
> uhm, because of that error, I cannot see the diagnostic provided by the linter `rust-analyzer`.
> 
> Inside cargo project: ![image](https://private-user-images.githubusercontent.com/57800056/371504254-2ef73f4e-7ada-45dc-8787-1f662a26f218.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc0MzYwMTQsIm5iZiI6MTcyNzQzNTcxNCwicGF0aCI6Ii81NzgwMDA1Ni8zNzE1MDQyNTQtMmVmNzNmNGUtN2FkYS00NWRjLTg3ODctMWY2NjJhMjZmMjE4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI3VDExMTUxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc2NDNkYTEzZTQxZmYxMGEwODZjODNlYjVmYWM0NTBlNmZlNmFjMmY1MTYxOWYwMDk2ODE0OWE3OTU3ZDMzNzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.T9hlhePXAkaL5LriqP2kRD5IvLsHr2IfhgYLXaUSqg8)
> 
> Standalone file: ![image](https://private-user-images.githubusercontent.com/57800056/371504358-7006af34-e539-4635-9210-955297a09ecc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc0MzYwMTQsIm5iZiI6MTcyNzQzNTcxNCwicGF0aCI6Ii81NzgwMDA1Ni8zNzE1MDQzNTgtNzAwNmFmMzQtZTUzOS00NjM1LTkyMTAtOTU1Mjk3YTA5ZWNjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI3VDExMTUxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ5MzczZDcxMmVjYmMzMGNhNjIwOGI1ZTI2OGMyNjc0MzE0MjFjODNkNWUxMGE5NGRjODY1OGE4Y2FhODQ0NWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.4GqqawGvo1cXIccHw6l0bEcsYTJuHdOBRq8FoHg_GfE)

rust-analyzer has limited support for standalone files, I'm afraid.

**Reply by barraIhsan** [reply]
> > ![image](https://private-user-images.githubusercontent.com/57800056/371503971-612a59c6-cb5d-425a-960e-d9a11684d227.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc0MzU5MjQsIm5iZiI6MTcyNzQzNTYyNCwicGF0aCI6Ii81NzgwMDA1Ni8zNzE1MDM5NzEtNjEyYTU5YzYtY2I1ZC00MjVhLTk2MGUtZDlhMTE2ODRkMjI3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI3VDExMTM0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWViOGZlNjVhYWI5YjlmMGExYWI2YTliMGJiYjViZjlhNzU4NmE1MjE2NmQ4YzA1MzcxNTU3MjJmN2Y4NGRiNTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.M0tw0qyVddLrOG8nRYQ3imvWyX3X3mqu50k6ml0VPaA) this also a bug from their end?
> 
> I don't know what that is, but it's coming from rust-analyzer.

yeah and sometimes it will just crash the lsp (no autocompletion, etc.)

**Reply by barraIhsan** [reply]
> > > If it bothers you, you can disable rust-analyzer server status notifications by setting the vim.g.rustaceanvim.server.status_notify_level option to false.
> > 
> > 
> > uhm, because of that error, I cannot see the diagnostic provided by the linter `rust-analyzer`.
> > Inside cargo project: ![image](https://private-user-images.githubusercontent.com/57800056/371504254-2ef73f4e-7ada-45dc-8787-1f662a26f218.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc0MzYwMTQsIm5iZiI6MTcyNzQzNTcxNCwicGF0aCI6Ii81NzgwMDA1Ni8zNzE1MDQyNTQtMmVmNzNmNGUtN2FkYS00NWRjLTg3ODctMWY2NjJhMjZmMjE4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI3VDExMTUxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc2NDNkYTEzZTQxZmYxMGEwODZjODNlYjVmYWM0NTBlNmZlNmFjMmY1MTYxOWYwMDk2ODE0OWE3OTU3ZDMzNzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.T9hlhePXAkaL5LriqP2kRD5IvLsHr2IfhgYLXaUSqg8)
> > Standalone file: ![image](https://private-user-images.githubusercontent.com/57800056/371504358-7006af34-e539-4635-9210-955297a09ecc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc0MzYwMTQsIm5iZiI6MTcyNzQzNTcxNCwicGF0aCI6Ii81NzgwMDA1Ni8zNzE1MDQzNTgtNzAwNmFmMzQtZTUzOS00NjM1LTkyMTAtOTU1Mjk3YTA5ZWNjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI3VDExMTUxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ5MzczZDcxMmVjYmMzMGNhNjIwOGI1ZTI2OGMyNjc0MzE0MjFjODNkNWUxMGE5NGRjODY1OGE4Y2FhODQ0NWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.4GqqawGvo1cXIccHw6l0bEcsYTJuHdOBRq8FoHg_GfE)
> 
> rust-analyzer has limited support for standalone files, I'm afraid.

oh, from the beginning? I just recently do `rustup update`

**Reply by mrcjkb** [reply]
For many diagnostics, rust-analyzer needs a workspace/cargo project (many diagnostics come from cargo).
If there's no cargo project, there's no cargo diagnostics.

**Reply by barraIhsan** [reply]
> For many diagnostics, rust-analyzer needs a workspace/cargo project (many diagnostics come from cargo). If there's no cargo project, there's no cargo diagnostics.

oh, ok then, Thanks!
I'll just do a simple `Cargo.toml` then
![image](https://github.com/user-attachments/assets/d3351ec9-847d-4f6f-bd9b-cd965c521fb6)
